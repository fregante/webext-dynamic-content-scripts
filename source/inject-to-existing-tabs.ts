import {injectContentScript} from 'webext-content-scripts';

export function injectToExistingTabs(
	origins: string[],
	scripts: ManifestContentScripts) {
	if (origins.length === 0) {
		return;
	}

	chrome.tabs.query({
		url: origins,
	}, tabs => {
		for (const tab of tabs) {
			if (tab.id) {
				void injectContentScript(tab.id, scripts);
			}
		}
	});
}
