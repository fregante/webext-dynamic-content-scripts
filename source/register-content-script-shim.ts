import {injectToExistingTabs} from './inject-to-existing-tabs.js';

type ManifestContentScript = NonNullable<chrome.runtime.Manifest['content_scripts']>[0];

// In Firefox, paths in the manifest are converted to full URLs under `moz-extension://` but browser.contentScripts expects exclusively relative paths
function makePathRelative(file: string): string {
	return new URL(file, location.origin).pathname;
}

export async function registerContentScript(id: string, origin: string, config: ManifestContentScript) {
	try {
		await chrome.scripting.registerContentScripts([{
			id,
			// Always convert paths here because we don't know whether Firefox MV3 will accept full URLs
			js: config.js?.map(file => makePathRelative(file)),
			css: config.css?.map(file => makePathRelative(file)),
			allFrames: config.all_frames,
			matches: [origin],
			excludeMatches: config.matches,
			persistAcrossSessions: true,
			runAt: config.run_at as browser.extensionTypes.RunAt,
		}]);
	} catch (error) {
		if ((error as Error)?.message.startsWith('Duplicate script ID')) {
			// Previously registered, nothing to do
			return;
		}

		// Unknown error, throw
		throw error;
	}

	// Registration successful, now inject into existing tabs
	await injectToExistingTabs([origin], [config]);
}
