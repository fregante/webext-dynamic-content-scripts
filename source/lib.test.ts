import {chrome} from 'jest-chrome';
import {describe, it, vi, beforeEach, expect} from 'vitest';
import {getAdditionalPermissions} from 'webext-additional-permissions';
import {init} from './lib.js';
import {injectToExistingTabs} from './inject-to-existing-tabs.js';
import {registerContentScript} from './register-content-script-shim.js';

vi.mock('webext-additional-permissions');
vi.mock('./register-content-script-shim.js');
vi.mock('./inject-to-existing-tabs.js');

const baseManifest: chrome.runtime.Manifest = {
	name: 'required',
	manifest_version: 3,
	version: '0.0.0',
	content_scripts: [
		{
			js: ['script.js'],
			matches: ['https://content-script.example.com/*'],
		},
	],
	host_permissions: ['https://permission-only.example.com/*'],
	optional_host_permissions: ['*://*/*'],
};

const additionalPermissions: Required<chrome.permissions.Permissions> = {
	origins: ['https://granted.example.com/*'],
	permissions: [],
};

const getAdditionalPermissionsMock = vi.mocked(getAdditionalPermissions);
const injectToExistingTabsMock = vi.mocked(injectToExistingTabs);
const registerContentScriptMock = vi.mocked(registerContentScript);

beforeEach(() => {
	registerContentScriptMock.mockClear();
	injectToExistingTabsMock.mockClear();
	getAdditionalPermissionsMock.mockResolvedValue(additionalPermissions);
	chrome.runtime.getManifest.mockReturnValue(baseManifest);
});

describe('init', () => {
	it('it should register the listeners and start checking permissions', async () => {
		await init();
		expect(getAdditionalPermissionsMock).toHaveBeenCalled();
		expect(injectToExistingTabsMock).toHaveBeenCalledWith(
			additionalPermissions.origins,
			baseManifest.content_scripts,
		);

		// TODO: https://github.com/extend-chrome/jest-chrome/issues/20
		// expect(chrome.permissions.onAdded.addListener).toHaveBeenCalledOnce();
		// expect(chrome.permissions.onRemoved.addListener).toHaveBeenCalledOnce();
	});

	it('it should throw if no content scripts exist at all', async () => {
		const manifest = structuredClone(baseManifest);
		delete manifest.content_scripts;
		chrome.runtime.getManifest.mockReturnValue(manifest);
		await expect(init()).rejects.toMatchInlineSnapshot('[Error: webext-dynamic-content-scripts tried to register scripts on the new host permissions, but no content scripts were found in the manifest.]');
	});
});

describe('init - registerContentScript', () => {
	it('should register the manifest scripts on new permissions', async () => {
		await init();
		expect(registerContentScriptMock).toMatchSnapshot();
	});

	it('should register the manifest scripts on multiple new permissions', async () => {
		getAdditionalPermissionsMock.mockResolvedValue({
			origins: [
				'https://granted.example.com/*',
				'https://granted-more.example.com/*',
			],
			permissions: [],
		});

		await init();
		expect(registerContentScriptMock).toMatchSnapshot();
	});

	it('should register multiple manifest scripts on new permissions', async () => {
		const manifest = structuredClone(baseManifest);
		manifest.content_scripts!.push({
			js: ['otherScript.js'],
			matches: ['https://content-script-extra.example.com/*'],
		});
		chrome.runtime.getManifest.mockReturnValue(manifest);

		await init();
		expect(registerContentScriptMock).toMatchSnapshot();
	});
});
