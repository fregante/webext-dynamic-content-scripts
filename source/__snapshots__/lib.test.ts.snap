// Vitest Snapshot v1

exports[`init - registerContentScript > should register multiple manifest scripts on new permissions 1`] = `
[MockFunction registerContentScript] {
  "calls": [
    [
      {
        "allFrames": undefined,
        "css": undefined,
        "excludeMatches": [
          "https://content-script.example.com/*",
        ],
        "js": [
          "/script.js",
        ],
        "matches": [
          "https://granted.example.com/*",
        ],
        "runAt": undefined,
      },
    ],
    [
      {
        "allFrames": undefined,
        "css": undefined,
        "excludeMatches": [
          "https://content-script-extra.example.com/*",
        ],
        "js": [
          "/otherScript.js",
        ],
        "matches": [
          "https://granted.example.com/*",
        ],
        "runAt": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init - registerContentScript > should register the manifest scripts on multiple new permissions 1`] = `
[MockFunction registerContentScript] {
  "calls": [
    [
      {
        "allFrames": undefined,
        "css": undefined,
        "excludeMatches": [
          "https://content-script.example.com/*",
        ],
        "js": [
          "/script.js",
        ],
        "matches": [
          "https://granted.example.com/*",
        ],
        "runAt": undefined,
      },
    ],
    [
      {
        "allFrames": undefined,
        "css": undefined,
        "excludeMatches": [
          "https://content-script.example.com/*",
        ],
        "js": [
          "/script.js",
        ],
        "matches": [
          "https://granted-more.example.com/*",
        ],
        "runAt": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`init - registerContentScript > should register the manifest scripts on new permissions 1`] = `
[MockFunction registerContentScript] {
  "calls": [
    [
      {
        "allFrames": undefined,
        "css": undefined,
        "excludeMatches": [
          "https://content-script.example.com/*",
        ],
        "js": [
          "/script.js",
        ],
        "matches": [
          "https://granted.example.com/*",
        ],
        "runAt": undefined,
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
